File: 396-rotate-function/rotate-function.dart
class Solution {
  int maxRotateFunction(List<int> nums) {
    int n = nums.length;

    int sum = 0;
    int f0 = 0;

    for (int i = 0; i < n; i++) {
      sum += nums[i];
      f0 += i * nums[i];
    }

    int maxVal = f0;
    int prev = f0;

    for (int k = 1; k < n; k++) {
      prev = prev + sum - n * nums[n - k];
      if (prev > maxVal) {
        maxVal = prev;
      }
    }

    return maxVal; // ✅ guaranteed return
  }
}

File: 396-rotate-function/rotate-function.py
class Solution(object):
    def maxRotateFunction(self, nums):
        n = len(nums)
        total_sum = sum(nums)

        # F(0)
        F = sum(i * num for i, num in enumerate(nums))
        max_val = F

        # Compute next rotations using formula
        for k in range(1, n):
            F = F + total_sum - n * nums[n - k]
            max_val = max(max_val, F)

        return max_val